src/main/java/jenkins/branch/NameMangler.java

```diff
@@ -0,0 +1,172 @@
+                switch (name.toLowerCase(Locale.ENGLISH)) {
+                    case ".":
+                    case "..":
+                    case "con":
+                    case "prn":
+                    case "aux":
+                    case "nul":
+                    case "com1":
+                    case "com2":
+                    case "com3":
+                    case "com4":
+                    case "com5":
+                    case "com6":
+                    case "com7":
+                    case "com8":
+                    case "com9":
+                    case "lpt1":
+                    case "lpt2":
+                    case "lpt3":
+                    case "lpt4":
+                    case "lpt5":
+                    case "lpt6":
+                    case "lpt7":
+                    case "lpt8":
+                    case "lpt9":
+                        unsafe = true;
+                        break;
+                    default:
+                        if (name.endsWith(".")) {
+                            unsafe = true;
+                        }
+                        break;
+                }
+            }
+            if (!unsafe) {
+                return name;
+            }
+        }
+        StringBuilder buf = new StringBuilder(name.length() + 16);
+        for (char c : name.toCharArray()) {
+            if (isSafe(c)) {
+                buf.append(c);
+            } else if (c == '/' || c == '\\' || c == ' ') {
+                buf.append('_');
+            } else if (c <= 0xff) {
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString(c & 0xff), 2, '0'));
+            } else {
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString(c & 0xff), 2, '0'));
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString((c & 0xffff) >> 8), 2, '0'));
+            }
+        }
+        // use the digest of the original name
+        String digest;
+        try {
+            MessageDigest sha = MessageDigest.getInstance("SHA-1");
+            byte[] bytes = sha.digest(name.getBytes(StandardCharsets.UTF_8));
+            int bits = 0;
+            int data = 0;
+            StringBuffer dd = new StringBuffer(32);
+            for (byte b : bytes) {
+                while (bits >= 5) {
+                    dd.append(toDigit(data & 0x1f));
+                    bits -= 5;
+                    data = data >> 5;
+                }
+                data = data | ((b & 0xff) << bits);
+                bits += 8;
+            }
+            digest = dd.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("MD5 not installed", e);    // impossible
+        }
+        if (buf.length() <= MAX_SAFE_LENGTH - MIN_HASH_LENGTH - 1) {
+            // we have room to add the min hash
+            buf.append('@');
```

Comment:
Could use

```java
System.getProperty(WorkspaceList.class.getName(), "@")
```

since apparently this character is problematic on some systems. OTOH `WorkspaceLocatorImpl` may already be handling that, according to the test—a little hard to follow.